build-support/native-image/build-zinc-native-image.bash

```diff
@@ -0,0 +1,120 @@
+# Functions to simplify the multiple complex bootstrapping techniques currently necessary to build a
+# zinc native-image.
+# TODO: This will be made automatic in pants via https://github.com/pantsbuild/pants/pull/6893.
+
+# TODO: build off of a more recent graal sha (more recent ones fail to build scalac and more): see
+# https://github.com/oracle/graal/issues/1448.
+
+# shellcheck source=build-support/native-image/utils.bash
+source "${SCRIPT_DIR}/utils.bash"
+
+# FUNCTIONS
+
+function _get_coursier_impl {
+  if [[ ! -f ./coursier ]]; then
+    curl -Lo coursier https://git.io/coursier-cli || return "$?"
+    chmod +x coursier
+    ./coursier --help >&2 || return "$?"
+  fi >&2
+  normalize_path coursier
+}
+
+function get_coursier {
+  do_within_cache_dir _get_coursier_impl
+}
+
+function bootstrap_environment {
+  # Install necessary tools for the ubuntu:latest container on docker hub.
+  if ! is_osx; then
+    apt-get update
+    apt-get -y install \
+            g{cc,++} git curl aptitude zlib1g-dev make python{,3} git python3-pip \
+            pkg-config libssl-dev libpython-dev openjdk-8-jdk
+
+    # FIXME: Otherwise pants will fail at bootstrap with an unrecognized symbol `distutils.spawn`.
+    pip3 install setuptools==40.0.0
+  fi
+  mkdir -pv "$NATIVE_IMAGE_BUILD_CACHE_DIR"
+}
+
+function get_base_native_image_build_script_graal_checkout {
+  do_within_cache_dir clone_repo_somewhat_idempotently \
+                      graal/ \
+                      https://github.com/cosmicexplorer/graal \
+                      graal-make-zinc-again
+}
+
+function clone_mx {
+  do_within_cache_dir clone_repo_somewhat_idempotently \
+                      mx/ \
+                      https://github.com/graalvm/mx \
+    && >&2 with_pushd "${NATIVE_IMAGE_BUILD_CACHE_DIR}/mx" \
+           ./mx update
+}
+
+function extract_openjdk_jvmci {
```

Comment:
Is this built-in to any modern JVMs, or is it definitely necessary to bootstrap it like this?

I think that I'm ok having this as a step in a script like this, but we're going to need to figure out an appropriate `binaryutil`/remoting split to 1) fetch, 2) use, this and the graal binary.

Reply:
This resource is necessary specifically to build graal (and more specifically to execute the `native-image` tool from source), because we're stuck on an old sha of it. If we were to make it a `NativeTool`, I would have this entire environment, *after* building the native image tool and the agent lib, made into a platform-specific `.tar.gz`. We could easily add in `mx`, `buildozer` ,and `jq` to that archive as well, or those could also all be their own snapshots.

Reply:
It's probably possible / easy to just "generate the graal release archive" from the graal repo. I should check the docs for their homegrown build tool.